src/features/ci-cd/hooks/release-versioning.ts

```diff
@@ -181,106 +180,100 @@ interface PatchCustomObject extends CustomObjectBase {
+	const filters: FilterObj[] = [];
+	if (custom.is_final) {
+		filters.push({
+			revision: { $eq: null },
+		});
+	}
+	if (custom.semver != null) {
+		filters.push({
+			revision: { $ne: null },
+			semver: { $ne: custom.semver },
+		});
+	}
+	if (request.values.belongs_to__application != null) {
+		filters.push({
+			revision: { $ne: null },
+			belongs_to__application: {
+				$ne: request.values.belongs_to__application,
 			},
-		})) as NonNullable<PatchCustomObject['releasesToSetRevision']>;
-		if (!releasesToSetRevision.length) {
-			return;
-		}
-		custom.releasesToSetRevision = releasesToSetRevision;
-
-		if (custom.semver == null) {
-			return;
-		}
-		// When changing the semver, set the revision to null so that
-		// we don't end up with duplicate revisions.
-		// We will set the correct value in PRERESPOND
-		await api.patch({
-			resource: 'release',
-			// Needs root because revision is not settable.
-			passthrough: { req: permissions.root },
-			options: {
-				$filter: {
-					id: { $in: releasesToSetRevision.map((r) => r.id) },
-					revision: { $ne: null },
-				},
+		});
+	}
+	if (filters.length === 0) {
+		// no field of interest was PATCHed
+		return;
+	}
+	const releaseIds = await sbvrUtils.getAffectedIds(args);
+	if (!releaseIds.length) {
+		return;
+	}
+	const releasesToSetRevision = (await api.get({
+		resource: 'release',
+		options: {
+			$select: [
+				'id',
+				'semver',
+				'is_finalized_at__date',
+				'belongs_to__application',
+			],
+			$filter: {
+				id: { $in: releaseIds },
+				...(filters.length === 1
+					? filters[0]
+					: {
+							$or: filters,
+					  }),
 			},
-			body: {
-				revision: null,
+			$orderby: [
+				// order first by application, so that the advisory locks are picked
+				// in a consistent order across requests so that we can avoid deadlocks
+				{ belongs_to__application: 'asc' },
+				// order by finalization date & id in order to have predictable revision ordering
+				{ is_finalized_at__date: 'asc' },
+				{ id: 'asc' },
+			],
+		},
+	})) as NonNullable<PatchCustomObject['releasesToSetRevision']>;
+	if (!releasesToSetRevision.length) {
+		return;
+	}
+	custom.releasesToSetRevision = releasesToSetRevision;
+
+	if (custom.semver == null) {
+		return;
+	}
+	// When changing the semver, set the revision to null so that
+	// we don't end up with duplicate revisions.
+	// We will set the correct value in PRERESPOND
+	await api.patch({
+		resource: 'release',
+		// Needs root because revision is not settable.
+		passthrough: { req: permissions.root },
+		options: {
+			$filter: {
+				id: { $in: releasesToSetRevision.map((r) => r.id) },
+				revision: { $ne: null },
 			},
-		});
+		},
+		body: {
+			revision: null,
+		},
+	});
+};
+
+hooks.addPureHook('PATCH', 'resin', 'release', {
+	POSTPARSE: parseReleaseVersioningFields,
+	PRERUN: async (args) => {
+		await Promise.all([
+			setReleasesToSetRevision(args),
+			preventChangingFinalToDraft(args),
```

Comment:
Do both of these rely on `parseReleaseVersioningFields`?

Reply:
`preventChangingFinalToDraft` depends on `parseReleaseVersioningFields` setting the `custom.is_final`.
In order to make `preventChangingFinalToDraft` to not rely on `parseReleaseVersioningFields`, we would have to move it to POSTPARSE and read the `values.is_final` directly, but then that would rely on the order that the hooks run, since `parseReleaseVersioningFields` also deletes the `values.is_final` since it doesn't exist in the DB.